File: scripts/generate-themes.py
import pystache
import json
import os

schemes_folder = 'temp/'
template = open('src/templates/bikkuri.mustache', 'r').read()
dist_folder = 'dist/'
theme_prefix = 'bikkuri-'


def delete_files(folder_path):
  files = os.listdir(folder_path)

  # iterate over the files and delete them
  for file_name in files:
    file_path = os.path.join(folder_path, file_name)
    try:
      if os.path.isfile(file_path):
        os.remove(file_path)
    except Exception as e:
      print(f"Failed to delete {file_path}. Reason: {e}")


def process_schemes(folder_path):
  # cleaning up dist folder
  delete_files(dist_folder)

  schemes_files = [f for f in os.listdir(folder_path) if f.endswith('.json')]

  themes_list = []  # for populating for package.json

  for scheme_file in schemes_files:
    scheme_filepath = os.path.join(folder_path, scheme_file)
    with open(scheme_filepath) as f:
      scheme_json = json.load(f)

      themes_list.append({
          "label":
          f"Bikkuri ({scheme_json['mode-name']} {scheme_json['tint-name']} & {scheme_json['mood-name']})",
          "uiTheme":
          f"{scheme_json['mode-type']}",
          "path":
          f"./{dist_folder + theme_prefix + scheme_file}"
      })

      render_theme(scheme_json, scheme_file)

  # add theme definitions to the package.json
  with open('package.json', 'r') as f:
    package_data = json.load(f)

  package_data["contributes"]["themes"] = themes_list

  with open('package.json', 'w') as f:
    json.dump(package_data, f, indent=2)

  print('Themes generated.')


def render_theme(scheme, filename):
  theme_filename = dist_folder + theme_prefix + filename
  with open(theme_filename, 'w') as f:
    f.write(pystache.render(template, scheme))
    f.close()


process_schemes(schemes_folder)

File: scripts/generate-themes.js
const fs = require('fs');
const path = require('path');

const schemesFolder = 'temp/';
const template = fs.readFileSync('src/templates/bikkuri.mustache', 'utf8');
const distFolder = 'dist/';
const themePrefix = 'bikkuri-';

function deleteFiles(folderPath) {
  if (!fs.existsSync(folderPath)) {
    fs.mkdirSync(folderPath, { recursive: true });
    return;
  }

  const files = fs.readdirSync(folderPath);

  files.forEach(fileName => {
    const filePath = path.join(folderPath, fileName);
    try {
      if (fs.lstatSync(filePath).isFile()) {
        fs.unlinkSync(filePath);
      }
    } catch (err) {
      console.error(`Failed to delete ${filePath}. Reason: ${err.message}`);
    }
  });
}

function renderMustache(template, data) {
  let result = template;
  
  result = result.replace(/\{\{\s*([^}]+)\s*\}\}/g, (match, key) => {
    const keys = key.trim().split('.');
    let value = data;
    
    for (const k of keys) {
      if (value && typeof value === 'object' && k in value) {
        value = value[k];
      } else {
        return match;
      }
    }
    
    return value !== undefined ? value : match;
  });
  
  return result;
}

function renderTheme(scheme, filename) {
  const themeFilename = path.join(distFolder, themePrefix + filename);

  try {
    const renderedTheme = renderMustache(template, scheme);
    fs.writeFileSync(themeFilename, renderedTheme);
  } catch (err) {
    console.error(`Error rendering theme ${filename}: ${err.message}`);
  }
}

function processSchemes(folderPath) {
  deleteFiles(distFolder);

  if (!fs.existsSync(folderPath)) {
    console.error(`Schemes folder ${folderPath} does not exist`);
    return;
  }

  const schemeFiles = fs.readdirSync(folderPath).filter(f => f.endsWith('.json'));
  const themesList = []; // for populating package.json

  schemeFiles.forEach(schemeFile => {
    const schemeFilepath = path.join(folderPath, schemeFile);
    
    try {
      const schemeData = fs.readFileSync(schemeFilepath, 'utf8');
      const schemeJson = JSON.parse(schemeData);

      themesList.push({
        "label": `Bikkuri (${schemeJson['mode-name']} ${schemeJson['tint-name']} & ${schemeJson['mood-name']})`,
        "uiTheme": `vs-${schemeJson['mode-type']}`,
        "path": `./${distFolder}${themePrefix}${schemeFile}`
      });

      renderTheme(schemeJson, schemeFile);
    } catch (err) {
      console.error(`Error processing ${schemeFile}: ${err.message}`);
    }
  });

  // add theme definitions to the package.json
  try {
    const packageData = JSON.parse(fs.readFileSync('package.json', 'utf8'));
    packageData.contributes.themes = themesList;
    fs.writeFileSync('package.json', JSON.stringify(packageData, null, 2));
  } catch (err) {
    console.error(`Error updating package.json: ${err.message}`);
    return;
  }

  console.log('Themes generated.');
}

processSchemes(schemesFolder);